### PR TITLE
PRIME-2088 Associate HA technical support with vendor(s)

### PR DESCRIPTION
**Pull request by neophyte57**
| Scenario                                                                                                                                       | Expected                                                                                                                    |
| ---------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------------------------------------- |
| For a pre-existing Health Authority, update the Tech Support details                                                                           | Update of Tech Support details is successful and persistent                                                                 |
| Enter the details of a new Health Authority (i.e. not updating the details of a pre-existing Health Authority), including Tech Support details | All Health Authority details are saved successfully                                                                         |
| Change which Vendors a Tech Support Contact is associated with                                                                                 | New associations are saved and \`HealthAuthorityTechnicalSupportVendor\` database table is cleaned of previous associations |
| Save with no Vendors supported                                                                                                                 | Save is successful                                                                                                          |
| 1\. Check some vendors                                                                                                                         | Other vendors become unchecked                                                                                              |
| 2\. Check "None"                                                                                                                               |
| 1\. Check "None"                                                                                                                               | "None" should become unchecked                                                                                              |
| 2\. Then check any other vendor                                                                                                                |
| 1\. Change vendors selected                                                                                                                    | Changed vendors should be reflected in UI                                                                                   |
| 2\. Save and Continue                                                                                                                          |
| 3\. Edit same Tech Support contact                                                                                                             |
| 1\. Edit more than 1 Tech Support contact such that each has different vendors supported                                                       | Vendors supported by each Tech Support contact should be accurate                                                           |
| 2\. After "Save and Continue", review vendors supported by each Tech Support contact                                                           |
| Update PharmaNet Administrator                                                                                                                 | Update is successful                                                                                                        |

### CODE REVIEW

File: prime-angular-frontend/src/app/lib/classes/technical-supports-form-state.spec.ts [hunk 1]
```diff
@@ -0,0 +1,7 @@
+import { TechnicalSupportsFormState } from './technical-supports-form-state';
+
+describe('TechnicalSupportsFormState', () => {
+  it('should create an instance', () => {
+    expect(new TechnicalSupportsFormState()).toBeTruthy();
```
Comment (johnnyl6) [inline prime-angular-frontend/src/app/lib/classes/technical-supports-form-state.spec.ts:5]: it seems that a couple of FE tests are failing
image

File: prime-angular-frontend/src/app/modules/adjudication/pages/health-authorities/technical-supports-page/technical-supports-page.component.ts [hunk 2]
```diff
@@ -37,16 +48,84 @@ export class TechnicalSupportsPageComponent extends AbstractContactsPage impleme
...
-  protected performSubmissionRequest(contact: Contact[]): NoContent {
-    return this.healthAuthResource.updateHealthAuthorityTechnicalSupports(this.route.snapshot.params.haid, contact);
+  protected getContactListItem(): (contact: HealthAuthorityTechnicalSupport) => CardListItem {
+    const healthAuthorityVendorPipe = new HealthAuthorityVendorPipe(this.configService);
+    return (contact: HealthAuthorityTechnicalSupport) => ({
+      icon: 'account_circle',
+      title: `${this.cardTitlePrefix}${contact.firstName} ${contact.lastName}`,
+      properties: [
+        { key: 'Job Title', value: contact.jobRoleTitle },
+        { key: 'Vendor', value: healthAuthorityVendorPipe.transform(contact.vendorsSupported) }
```
Comment (hogmann) [inline prime-angular-frontend/src/app/modules/adjudication/pages/health-authorities/technical-supports-page/technical-supports-page.component.ts:124]: Might want this to be "Vendor(s)"

File: prime-angular-frontend/src/app/lib/classes/abstract-contacts-page.class.ts [hunk 3]
```diff
@@ -187,6 +187,15 @@ export abstract class AbstractContactsPage extends AbstractEnrolmentPage {
    */
   protected abstract performSubmissionRequest(contact: Contact[]): NoContent;
 
+  /**
+   * @description
+   * Hook to capture any non-contact details, for example.  Returns post-manipulation JSON
+   */
+  protected manipulateFormStateJsonPreSubmission(json: any): any {
```
Comment (hogmann) [inline prime-angular-frontend/src/app/lib/classes/abstract-contacts-page.class.ts:194]: Could probably remove FormState to shorten this a little.

File: prime-angular-frontend/src/app/modules/adjudication/pages/health-authorities/technical-supports-page/technical-supports-page.component.html [hunk 4]
```diff
@@ -13,6 +13,30 @@
         <ng-container appPageSubheader2Summary>
           Provide information for a Technical Support Contact for this Health Authority.
         </ng-container>
+
+        <ng-container nonContactDetails>
+          <app-page-subheader2>
+            <ng-container appPageSubheader2Title>Vendor</ng-container>
+            <ng-container appPageSubheader2Summary>
+              What vendor does this PharmaNet Administrator work with?
```
Comment (hogmann) [inline prime-angular-frontend/src/app/modules/adjudication/pages/health-authorities/technical-supports-page/technical-supports-page.component.html:21]: Vendor(s), hopefully the language still works

**Review by hogmann — APPROVED** [review]
Other than some small text changes everything looks good 👍

File: prime-angular-frontend/src/app/shared/components/site/party-review/party-review.component.html [hunk 5]
```diff
@@ -25,6 +25,11 @@
   {{ party?.email | default }}
 </app-enrollee-property>
 
+<app-enrollee-property *ngIf="$any(party).vendorsSupported !== undefined"
+                       title="Vendor">
+  {{ $any(party).vendorsSupported | healthAuthorityVendor | default }}
```
Comment (johnnyl6) [inline prime-angular-frontend/src/app/shared/components/site/party-review/party-review.component.html:30]: is there any work-around for the `$any()` type casting here? It looks like the input `[party]` is type `Party` but an arbitrary `HealthAuthorityTechnicalSupport` is passed and force casted to check `vendorsSupported`. feels like the input `party` is *polluted* here.  could the the force-casting be eliminated by pass in a separate input param like `healthAutorityTechnicalSupport`?

Comment (neophyte57) [reply prime-angular-frontend/src/app/shared/components/site/party-review/party-review.component.html:30]: Good point ... I'm trying an alternative to `$any()` but so far not working 

image

Did you have something in mind @johnnyl6 ?

Comment (johnnyl6) [reply prime-angular-frontend/src/app/shared/components/site/party-review/party-review.component.html:30]: I was thinking of a separate input param to hold the `vendorSupported` property in `PartyReviewComponent`
```typescript
@Input() public healthAutorityTechnicalSupport: HealthAuthorityTechnicalSupport;
```
```html
<app-enrollee-property *ngIf="healthAutorityTechnicalSupport?.vendorsSupported"
                       title="Vendor">
  {{ healthAutorityTechnicalSupport.vendorsSupported | healthAuthorityVendor | default }}
</app-enrollee-property>
```

Comment (hogmann) [reply prime-angular-frontend/src/app/shared/components/site/party-review/party-review.component.html]: Sorry to jump in here, but this is another place where projection makes it easier. and doesn't cause us to have to add outside knowledge to a generic component. 

image

So by doing something as simple as a content select on app-enrollee-properties inside the party-review.component.ts. Now whenever we use the component we can wrap additional properties and they are displayed before the address. So then in the health-auth-org-info-page.component, which has more context and knows of the HealthAuthTechSupport you can just do something like this

image

As you can see i pass in multiple to show it doesn't limit you to one. Now the shared generic component doesn't have to have the context from it's parent to be able to render this new field.

**Review by johnnyl6 — APPROVED** [review]
LGTM

**Review by hogmann — APPROVED** [review]

File: prime-angular-frontend/src/app/shared/components/site/party-review/party-review.component.html (hunk 5 above)
Comment (neophyte57) [reply prime-angular-frontend/src/app/shared/components/site/party-review/party-review.component.html]: @hogmann Great suggestion!

**Pull request merged**